tests/agent/nanojava_interpreter_test.cc

```diff
@@ -0,0 +1,3035 @@
+BINARY_ARITHMETIC_OPERATION_TEST(l, rem, %, 436727828L, 23735L)
+BINARY_ARITHMETIC_OPERATION_TEST(l, and, &, 348953897234L, 23745264563345L)
+BINARY_ARITHMETIC_OPERATION_TEST(l, or, |, 348953897234L, 23745264563345L)
+BINARY_ARITHMETIC_OPERATION_TEST(l, xor, ^, 348953897234L, 23745264563345L)
+BINARY_ARITHMETIC_OPERATION_TEST(d, add, +, 345.342345, 223434.234234)
+BINARY_ARITHMETIC_OPERATION_TEST(d, sub, -, -23482349723.554, 891282346.123)
+BINARY_ARITHMETIC_OPERATION_TEST(d, mul, *, 2323.45, 128.3487232344)
+BINARY_ARITHMETIC_OPERATION_TEST(d, div, /, 7177.8743452, 4.3221343)
+
+
+TEST_F(NanoJavaInterpreterTest, IntegerDivisionOverflow) {
+  const std::string test_cases[] = {"idiv", "irem"};
+  for (const std::string& test_case : test_cases) {
+    const std::string blob = AssembleMethod("I", Substitute(
+                                                     R"(.limit stack 2
+               ldc -2147483648
+               ldc -1
+               $0
+               ireturn)",
+                                                     test_case));
+
+    ASSERT_EQ("<int>-2147483648", ExecuteExpectSuccess(blob, {}));
+  }
+}
+
+
+TEST_F(NanoJavaInterpreterTest, LongDivisionOverflow) {
+  const std::string test_cases[] = {"ldiv", "lrem"};
+  for (const std::string& test_case : test_cases) {
+    const std::string blob = AssembleMethod("J", Substitute(
+                                                     R"(.limit stack 4
+               ldc2_w -9223372036854775808
+               ldc2_w -1
+               $0
+               lreturn)",
+                                                     test_case));
+
+    ASSERT_EQ("<long>-9223372036854775808", ExecuteExpectSuccess(blob, {}));
+  }
+}
+
+
+TEST_F(NanoJavaInterpreterTest, IntegerDivisionByZero) {
+  const std::string test_cases[] = {"idiv", "irem"};
+  for (const std::string& test_case : test_cases) {
+    const std::string blob = AssembleMethod("I", Substitute(
+                                                     R"(.limit stack 2
+               ldc 47
+               ldc 0
+               $0
+               ireturn)",
+                                                     test_case));
+
+    ASSERT_EQ(
+        "java.lang.ArithmeticException",
+        ExecuteExpectException(blob, {}));
+  }
+}
+
+
+TEST_F(NanoJavaInterpreterTest, LongDivisionByZero) {
+  const std::string test_cases[] = {"ldiv", "lrem"};
+  for (const std::string& test_case : test_cases) {
+    const std::string blob = AssembleMethod("J", Substitute(
+                                                     R"(.limit stack 4
+               ldc2_w 472345723423432
+               ldc2_w 0
+               $0
+               lreturn)",
+                                                     test_case));
+
+    ASSERT_EQ(
+        "java.lang.ArithmeticException",
+        ExecuteExpectException(blob, {}));
+  }
+}
+
+
+// static int test(int x) {
+//   return -x;
+// }
+TEST_F(NanoJavaInterpreterTest, Opcode_ineg) {
```

Comment:
Is there any guidance available for how to reconstruct these tests/add to them if needed?

Reply:
Not that I'm aware of.